### DISCUSSION TITLE
Load datapacks

### DISCUSSION THREAD

**Discussion by wiligl** [itzg/docker-minecraft-server#1988] [Q&A]
Hi, I am running the following itzg/minecraft-server on a QNAP TS-653A (QTS5) NAS using docker:

docker run -d \
           -e MEMORY=2G \
           -p 25565:25565 \
           -e EULA=TRUE \
           -v /share/homes/wiligl/minecraft/MC_V1.19.3_DP/:/data \
           -v /share/homes/wiligl/minecraft/MC_V1.19.3_DP/world/datapacks/:/datapacks \
           --restart always \
           --name NAS_MC_V1.19.3_DP itzg/minecraft-server

I have copied three datapacks (eg life-steal-free-e700.zip, bucketable-v2-3-data-pack.zip, galaxium-tools-armor-dp.zip) into the /datapacks/ folder on the host system, but if I start the server the datapacks are not found or loaded. Minecraft works but only as a standard VANILLA server.

The commands /reload or /datapacks list are not avaible.

How can I load and enable the datapacks?

Kr, Will

**Comment by itzg** [comment]
`/datapacks` isn't accessed automatically like some of the other directories mentioned in the docs. Instead you'll need to explicitly reference the ones you want loaded from that path:

https://github.com/itzg/docker-minecraft-server#datapacks

**Reply by wiligl** [reply]
You mean adding this line?
-e DATAPACKS=life-steal-free-e700.zip, bucketable-v2-3-data-pack.zip, galaxium-tools-armor-dp.zip

**Reply by itzg** [reply]
Yes but will need to be fully qualified paths like `/datapacks/life-steal-free-e700.zip `. Be sure to leave the volume attachment as you have it.

**Reply by wiligl** [reply]
This code gives errors:

docker run -d \
           -e MEMORY=2G \
           -p 25565:25565 \
           -e EULA=TRUE \
           -e MODE=creative \
           -v /share/homes/wiligl/minecraft/MC_V1.19.3_DP/:/data \
           -v /share/homes/wiligl/minecraft/MC_V1.19.3_DP/world/datapacks/:/datapacks \
           -e DATAPACKS=/datapacks/life-steal-free-e700.zip,/datapacks/bucketable-v2-3-data-pack.zip,/datapacks/galaxium-tools-armor-dp.zip \
           --restart always \
           --name NAS_MC_V1.19.3_DP itzg/minecraft-server
           
*Errors*:
[init] Running as uid=1000 gid=1000 with /data as 'drwxrwxr-x 8 1000 1000 4096 Mar  4 15:11 /data'
[init] Resolved version given LATEST into 1.19.3 and major version 1.19
[init] Resolving type given VANILLA
[init] Copying datapack located at /datapacks/life-steal-free-e700.zip ...
cp: '/datapacks/life-steal-free-e700.zip' and '/data/world/datapacks/life-steal-free-e700.zip' are the same file
[init] ERROR: failed to copy from /datapacks/life-steal-free-e700.zip into /data/world/datapacks

**Reply by itzg** [reply]
Sorry, I just now noticed your datapacks mount is already pointing as the resulting datapacks works directory.

Let's back up a bit, where do you have those datapack zips? You need to mount whatever directory that is.

**Comment by wiligl** [comment]
Thanks, I tried this code:

docker run -d \
           -e MEMORY=2G \
           -p 25565:25565 \
           -e EULA=TRUE \
           -e MODE=creative \
           -e DATAPACKS=life-steal-free-e700.zip,bucketable-v2-3-data-pack.zip,galaxium-tools-armor-dp.zip \
           -v /share/homes/wiligl/minecraft/mc_v1.19.3_dp_repo/datapacks/:/data/world/datapacks/ \
           -v /share/homes/wiligl/minecraft/mc_v1.19.3_dp/:/data/ \
           --restart always \
           --name mc_v1.19.3_dp itzg/minecraft-server

Note: I realized that I previously mounted the /data and the /datapacks folder to the same host folder (overloading), and copyied the datapacks folder leading to error that same files exists,
           
 Error:
 [init] ERROR Invalid URL or path given in DATAPACKS: life-steal-free-e700.zip

Paths to datapack files in host folder:
[~] # ls -la /share/homes/wiligl/minecraft/mc_v1.19.3_dp_repo/datapacks/      
total 56
drwxrwxrwx 2 admin administrators  4096 2023-03-05 15:23 ./
drwxr-xr-x 3 admin administrators  4096 2023-03-05 15:57 ../
-rw-rw-rw- 1 admin administrators 16781 2023-02-11 09:58 bucketable-v2-3-data-pack.zip
-rw-rw-rw- 1 admin administrators 13842 2023-02-11 09:59 galaxium-tools-armor-dp.zip
-rw-rw-rw- 1 admin administrators  9156 2023-03-02 20:23 life-steal-free-e700.zip

Do I have to change file permissions? I am running docker as admin, so I would have guessed that this should work fine.


 
 How do I fix this ?

**Reply by wiligl** [reply]
Strangly, even if I ssh into the NAS as admin and run "docker", it creates the /data  folder on the host automatically with owner "wiligl".

**Reply by itzg** [reply]
OK, so you already have the datapacks manually placed into your world data -- you don't need the `DATAPACKS` feature. 

Just remove

```
-e DATAPACKS=life-steal-free-e700.zip,bucketable-v2-3-data-pack.zip,galaxium-tools-armor-dp.zip
-v /share/homes/wiligl/minecraft/mc_v1.19.3_dp_repo/datapacks/:/data/world/datapacks/
```

**Comment by wiligl** [comment]
Ok, so long story short if I run the docker minecraft server first to create the folder structure and manually copy the datapacks zip files into the host folder ./world/datapacks/, they will be loaded automagically and additional DATAPACKS volumes or flags will cause errors. Please add this to the documentation!
Note: Fun thing is that I tried this approach at the very beginning by copying the files into the /Container/container-station-data/lib/docker/volumes/d67e3fc942c1d...90de91/_data/world/datapacks/
folder automatically created by the QNAP/QTS/ContainerStation which did not work and datapacks were just ignored.
 (*sobbing quietly*)

**Reply by wiligl** [reply]
Datapacks are loaded now, but one throws an error!

---LOG---
[16:10:27] [ServerMain/INFO]: Building unoptimized datafixer
[16:10:34] [ServerMain/INFO]: Environment: authHost='https://authserver.mojang.com', accountsHost='https://api.mojang.com', sessionHost='https://sessionserver.mojang.com', servicesHost='https://api.minecraftservices.com', name='PROD'
[16:10:34] [ServerMain/INFO]: Found new data pack file/bucketable-v2-3-data-pack.zip, loading it automatically
[16:10:34] [ServerMain/INFO]: Found new data pack file/galaxium-tools-armor-dp.zip, loading it automatically
[16:10:34] [ServerMain/INFO]: Found new data pack file/life-steal-free-e700.zip, loading it automatically
[16:10:47] [ServerMain/INFO]: Loaded 7 recipes
[16:10:47] [ServerMain/ERROR]: Failed to load function lifesteal:death_ban
java.util.concurrent.CompletionException: java.lang.IllegalArgumentException: Whilst parsing command on line 1: Unknown or incomplete command, see below for error at position 0: <--[HERE]
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$AsyncSupply.exec(Unknown Source) ~[?:?]
	at java.util.concurrent.ForkJoinTask.doExec(Unknown Source) ~[?:?]
	at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source) ~[?:?]
	at java.util.concurrent.ForkJoinPool.scan(Unknown Source) ~[?:?]
	at java.util.concurrent.ForkJoinPool.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source) ~[?:?]
Caused by: java.lang.IllegalArgumentException: Whilst parsing command on line 1: Unknown or incomplete command, see below for error at position 0: <--[HERE]
	at dj.a(SourceFile:63) ~[server-1.19.3.jar:?]
	at acs.a(SourceFile:83) ~[server-1.19.3.jar:?]
	... 7 more
[16:10:47] [ServerMain/ERROR]: Couldn't load tag minecraft:tick as it is missing following references: crafting_template:loop (from file/galaxium-tools-armor-dp.zip)
[16:10:48] [ServerMain/INFO]: Loaded 1198 advancements
[16:10:51] [Server thread/INFO]: Starting minecraft server version 1.19.3
....

**Reply by itzg** [reply]
I'll see if I can make the documentation any clearer.

As for that error, that's something to report to the datapack author.

**Comment by wiligl** [comment]
Ok, so finally I have successfully started the minecraft docker image and enabled the datapacks with the following command:
sudo docker run -d \
           -e EULA=TRUE \
           -e MEMORY=2G \
           -e MODE=creative \
           -e OPS=PutYourMinecraftNameHere \
           -p 25565:25565 \
           -v /share/homes/admin/minecraft/mc_v1.19.3_dp/:/data \
           --restart always \
           --name mc_v1.19.3_dp itzg/minecraft-server
 
 Note: Every line should end with a backslash for the command line to see it as one command. The backslashes are not displayed here for some reason.
           
  I had to set myself to OPerator/admin to be able to access the /datapack commands.
  
  Thanks! You are hero for providing this docker image!! Great job!
